File: src/main/java/fisheye270/visualization/Mosaic.java
package fisheye270.visualization;

import fisheye270.runtime.PipelineResult;
import fisheye270.runtime.Processor;
import org.opencv.core.Core;
import org.opencv.core.CvType;
import org.opencv.core.Mat;
import org.opencv.core.Point;
import org.opencv.core.Scalar;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;

import java.util.ArrayList;
import java.util.List;

public final class Mosaic {
    private Mosaic() {}

    public static Mat view(Processor proc, PipelineResult r, String view) {
        return switch (view) {
            case "undistorted" -> proc.labeledQuad(r.undistorted, "View 2 - PERSPECTIVE UNDISTORT (debug)");
            case "projected" -> proc.labeledQuad(r.projected, "View 3 - " + proc.spec.model.toUpperCase());
            case "overlap" -> overlap(r);
            case "stitch" -> caption(orEmpty(r.panorama),
                    "View 7 - STITCH fov=" + (int) proc.spec.fovHDeg + " missing=" +
                            (r.missing.isEmpty() ? "none" : String.join(",", r.missing)));
            case "seams" -> caption(orEmpty(r.seams), "View 8 - SEAM WINNER");
            case "compare_proj" -> compare(proc, r);
            default -> proc.labeledQuad(r.raw, "View 1 - RAW");
        };
    }

    static Mat overlap(PipelineResult r) {
        List<Mat> rows = new ArrayList<>();
        String[][] pairs = {{"LEFT", "FRONT"}, {"FRONT", "RIGHT"}};
        for (String[] p : pairs) {
            Mat a = r.projected.get(p[0]);
            Mat b = r.projected.get(p[1]);
            if (a == null || b == null) continue;
            Mat blend = new Mat();
            Core.addWeighted(a, 0.5, b, 0.5, 0, blend);
            Mat small = new Mat();
            Imgproc.resize(blend, small, new Size(960, 270));
            Imgproc.putText(small, "View 4 - overlap " + p[0] + " + " + p[1],
                    new Point(10, 28), Imgproc.FONT_HERSHEY_SIMPLEX, 0.7, new Scalar(255, 255, 255), 2);
            rows.add(small);
        }
        if (rows.isEmpty()) return Mat.zeros(270, 960, CvType.CV_8UC3);
        Mat out = new Mat();
        Core.vconcat(rows, out);
        return out;
    }

    static Mat compare(Processor proc, PipelineResult r) {
        Mat raw = r.raw.get("FRONT");
        if (raw == null) return Mat.zeros(240, 720, CvType.CV_8UC3);
        List<Mat> panels = new ArrayList<>();
        String saved = proc.spec.model;
        for (String name : new String[]{"perspective", "cylindrical", "spherical"}) {
            Mat img;
            if ("perspective".equals(name)) {
                img = r.undistorted.getOrDefault("FRONT", raw);
            } else {
                proc.spec.model = name;
                proc.invalidate("FRONT");
                img = proc.project("FRONT", raw)[0];
            }
            Mat small = new Mat();
            Imgproc.resize(img, small, new Size(640, 180));
            Imgproc.putText(small, name, new Point(10, 28),
                    Imgproc.FONT_HERSHEY_SIMPLEX, 0.7, new Scalar(255, 255, 255), 2);
            panels.add(small);
        }
        proc.spec.model = saved;
        proc.invalidate("FRONT");
        Mat out = new Mat();
        Core.vconcat(panels, out);
        return out;
    }

    static Mat orEmpty(Mat m) {
        return m == null || m.empty() ? Mat.zeros(200, 400, CvType.CV_8UC3) : m;
    }

    static Mat caption(Mat img, String text) {
        Mat out = img.clone();
        Imgproc.putText(out, text, new Point(10, 28),
                Imgproc.FONT_HERSHEY_SIMPLEX, 0.8, new Scalar(0, 0, 0), 4);
        Imgproc.putText(out, text, new Point(10, 28),
                Imgproc.FONT_HERSHEY_SIMPLEX, 0.8, new Scalar(255, 255, 255), 2);
        return out;
    }
}

File: fisheye_270/visualization/mosaic.py
"""Build debug mosaics for each development view (no GUI required)."""

from __future__ import annotations

from typing import Dict, Optional

import cv2
import numpy as np

from ..runtime.processor import PipelineResult, Processor
from ..stitching.blend import CAMERA_COLORS


def _caption(img: np.ndarray, text: str) -> np.ndarray:
    out = img.copy()
    cv2.putText(out, text, (10, 28), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 0, 0), 4)
    cv2.putText(out, text, (10, 28), cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 255, 255), 2)
    return out


def side_by_side(a: np.ndarray, b: np.ndarray, la: str, lb: str) -> np.ndarray:
    def fit(im):
        h, w = im.shape[:2]
        scale = 480 / max(h, 1)
        return cv2.resize(im, (int(w * scale), 480))
    a, b = fit(a), fit(b)
    return np.hstack([_caption(a, la), _caption(b, lb)])


def mosaic_from_result(proc: Processor, result: PipelineResult, view: str) -> np.ndarray:
    if view == "raw":
        return _caption(proc.label_quad(result.raw), "View 1 — RAW")
    if view == "undistorted":
        return _caption(proc.label_quad(result.undistorted), "View 2 — PERSPECTIVE UNDISTORT (debug)")
    if view == "projected":
        tiles = {k: v for k, v in result.projected.items()}
        return _caption(proc.label_quad(tiles), f"View 3 — {proc.spec.model.upper()} PROJECTION")
    if view == "overlap":
        return overlap_view(result)
    if view == "stitch":
        img = result.panorama if result.panorama is not None else np.zeros((200, 400, 3), np.uint8)
        miss = ",".join(result.missing) or "none"
        return _caption(img, f"View 7 — STITCH fov={proc.spec.fov_h_deg:.0f} missing={miss}")
    if view == "seams":
        img = result.seams if result.seams is not None else np.zeros((200, 400, 3), np.uint8)
        return _caption(img, "View 8 — SEAM WINNER")
    if view == "compare_proj":
        return compare_projections(proc, result)
    return proc.label_quad(result.raw)


def overlap_view(result: PipelineResult) -> np.ndarray:
    pairs = [("LEFT", "FRONT"), ("FRONT", "RIGHT")]
    rows = []
    for a, b in pairs:
        ia, ib = result.projected.get(a), result.projected.get(b)
        if ia is None or ib is None:
            continue
        blend = cv2.addWeighted(ia, 0.5, ib, 0.5, 0)
        tint = blend.copy()
        tint[:, :, 1] = cv2.addWeighted(ia[:, :, 1], 0.7, tint[:, :, 1], 0.3, 0)
        rows.append(_caption(cv2.resize(tint, (960, 270)), f"View 4 — overlap {a} + {b}"))
    if not rows:
        return np.zeros((270, 960, 3), np.uint8)
    return np.vstack(rows)


def compare_projections(proc: Processor, result: PipelineResult) -> np.ndarray:
    cid = "FRONT"
    raw = result.raw.get(cid)
    if raw is None:
        return np.zeros((240, 720, 3), np.uint8)
    panels = []
    for name in ("perspective", "cylindrical", "spherical"):
        if name == "perspective":
            img = result.undistorted.get(cid, raw)
        else:
            img, _ = proc.project(cid, raw, model_name=name)
        small = cv2.resize(img, (640, 180))
        panels.append(_caption(small, name))
    return np.vstack(panels)


def legend() -> np.ndarray:
    bar = np.zeros((40, 640, 3), np.uint8)
    x = 10
    for name, col in CAMERA_COLORS.items():
        cv2.rectangle(bar, (x, 8), (x + 24, 32), col, -1)
        cv2.putText(bar, name, (x + 30, 28), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
        x += 150
    return bar
